File: ios/App/Sources/Core/Domain/NavCaddy/Memory/MissPatternAggregator.swift
import Foundation

/// Aggregates miss patterns from shot history with rolling windows and decay.
///
/// Spec R5: Aggregates patterns by club, lie, and pressure context with time-based decay.
/// Q5: Rolling window with 14-day decay half-life.
@MainActor
final class MissPatternAggregator {
    private let repository: NavCaddyRepository
    private let decayCalculator: PatternDecayCalculator

    /// Default rolling window in days (spec: last 30 days)
    static let defaultWindowDays = 30

    /// Minimum shots required to form a pattern
    static let minimumShotsForPattern = 3

    /// Minimum confidence threshold for pattern inclusion
    static let minimumConfidence = 0.1

    init(
        repository: NavCaddyRepository,
        decayCalculator: PatternDecayCalculator = .default
    ) {
        self.repository = repository
        self.decayCalculator = decayCalculator
    }

    // MARK: - Pattern Aggregation

    /// Aggregates miss patterns from a collection of shots.
    ///
    /// - Parameters:
    ///   - shots: Array of shots to analyze
    ///   - applyDecay: Whether to apply time-based decay to confidence (default: true)
    /// - Returns: Array of aggregated miss patterns sorted by confidence
    func aggregatePatterns(from shots: [Shot], applyDecay: Bool = true) -> [MissPattern] {
        // Filter to only shots with miss directions (excluding straight)
        let missedShots = shots.filter { shot in
            guard let direction = shot.missDirection else { return false }
            return direction != .straight
        }

        // Group by direction
        let groupedByDirection = Dictionary(grouping: missedShots) { $0.missDirection! }

        var patterns: [MissPattern] = []

        for (direction, directionShots) in groupedByDirection {
            // Overall pattern for this direction
            if let pattern = createPattern(
                direction: direction,
                shots: directionShots,
                club: nil,
                pressureContext: nil,
                applyDecay: applyDecay
            ) {
                patterns.append(pattern)
            }

            // Patterns by club
            let byClub = Dictionary(grouping: directionShots) { $0.club.id }
            for (clubId, clubShots) in byClub {
                if let pattern = createPattern(
                    direction: direction,
                    shots: clubShots,
                    club: clubShots.first?.club,
                    pressureContext: nil,
                    applyDecay: applyDecay
                ) {
                    patterns.append(pattern)
                }
            }

            // Patterns under pressure
            let pressureShots = directionShots.filter { $0.pressureContext.hasPressure }
            if !pressureShots.isEmpty {
                if let pattern = createPattern(
                    direction: direction,
                    shots: pressureShots,
                    club: nil,
                    pressureContext: pressureShots.first?.pressureContext,
                    applyDecay: applyDecay
                ) {
                    patterns.append(pattern)
                }
            }
        }

        // Sort by confidence descending
        return patterns
            .filter { $0.confidence >= Self.minimumConfidence }
            .sorted { $0.confidence > $1.confidence }
    }

    /// Gets patterns for a specific club with decay applied.
    ///
    /// - Parameter clubId: The club ID to filter by
    /// - Returns: Array of miss patterns for the club
    func getPatternsForClub(_ clubId: String) async throws -> [MissPattern] {
        // Get recent shots for the club
        let shots = try await repository.getRecentShots(days: Self.defaultWindowDays)
        let clubShots = shots.filter { $0.club.id == clubId }

        return aggregatePatterns(from: clubShots, applyDecay: true)
    }

    /// Gets patterns for a specific pressure context.
    ///
    /// - Parameter context: The pressure context to filter by
    /// - Returns: Array of miss patterns under pressure
    func getPatternsForContext(_ context: PressureContext) async throws -> [MissPattern] {
        guard context.hasPressure else {
            return []
        }

        let pressureShots = try await repository.getShotsWithPressure()
        let recentShots = pressureShots.filter { shot in
            let daysAgo = Date().timeIntervalSince(shot.timestamp) / (24 * 60 * 60)
            return daysAgo <= Double(Self.defaultWindowDays)
        }

        return aggregatePatterns(from: recentShots, applyDecay: true)
    }

    /// Gets patterns filtered by miss direction.
    ///
    /// - Parameter direction: The miss direction to filter by
    /// - Returns: Array of patterns with the specified direction
    func getPatternsByDirection(_ direction: MissDirection) async throws -> [MissPattern] {
        let shots = try await repository.getRecentShots(days: Self.defaultWindowDays)
        let directionShots = shots.filter { $0.missDirection == direction }

        return aggregatePatterns(from: directionShots, applyDecay: true)
    }

    /// Gets all recent patterns across all contexts.
    ///
    /// - Parameter windowDays: Rolling window in days (default: 30)
    /// - Returns: Array of all miss patterns
    func getAllPatterns(windowDays: Int = defaultWindowDays) async throws -> [MissPattern] {
        let shots = try await repository.getRecentShots(days: windowDays)
        return aggregatePatterns(from: shots, applyDecay: true)
    }

    // MARK: - Private Helpers

    /// Creates a miss pattern from a group of shots.
    private func createPattern(
        direction: MissDirection,
        shots: [Shot],
        club: Club?,
        pressureContext: PressureContext?,
        applyDecay: Bool
    ) -> MissPattern? {
        guard shots.count >= Self.minimumShotsForPattern else {
            return nil
        }

        let frequency = shots.count
        let lastOccurrence = shots.map(\.timestamp).max() ?? Date()

        // Calculate base confidence from frequency
        // More shots = higher confidence, with diminishing returns
        let baseConfidence = min(1.0, Double(frequency) / 10.0)

        // Apply decay if requested
        let confidence = applyDecay
            ? baseConfidence * decayCalculator.calculateDecay(for: lastOccurrence)
            : baseConfidence

        return MissPattern(
            direction: direction,
            club: club,
            frequency: frequency,
            confidence: confidence,
            pressureContext: pressureContext,
            lastOccurrence: lastOccurrence
        )
    }
}

// MARK: - Pattern Statistics

extension MissPatternAggregator {
    /// Statistical summary of miss patterns.
    struct PatternStatistics {
        let totalShots: Int
        let totalMisses: Int
        let mostCommonDirection: MissDirection?
        let averageConfidence: Double
        let patternsFound: Int
    }

    /// Calculates statistics for a set of shots.
    ///
    /// - Parameter shots: Array of shots to analyze
    /// - Returns: Statistical summary
    func calculateStatistics(from shots: [Shot]) -> PatternStatistics {
        let patterns = aggregatePatterns(from: shots, applyDecay: true)

        let missedShots = shots.filter { shot in
            guard let direction = shot.missDirection else { return false }
            return direction != .straight
        }

        let directionFrequencies = Dictionary(grouping: missedShots) { $0.missDirection! }
            .mapValues { $0.count }

        let mostCommon = directionFrequencies.max { $0.value < $1.value }?.key

        let avgConfidence = patterns.isEmpty
            ? 0.0
            : patterns.map(\.confidence).reduce(0, +) / Double(patterns.count)

        return PatternStatistics(
            totalShots: shots.count,
            totalMisses: missedShots.count,
            mostCommonDirection: mostCommon,
            averageConfidence: avgConfidence,
            patternsFound: patterns.count
        )
    }
}

File: android/app/src/main/java/com/example/app/domain/navcaddy/memory/MissPatternAggregator.kt
package caddypro.domain.navcaddy.memory

import caddypro.data.navcaddy.repository.NavCaddyRepository
import caddypro.domain.navcaddy.models.Club
import caddypro.domain.navcaddy.models.MissDirection
import caddypro.domain.navcaddy.models.MissPattern
import caddypro.domain.navcaddy.models.PressureContext
import caddypro.domain.navcaddy.models.Shot
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Aggregates miss patterns from shot history.
 *
 * Analyzes shots within a rolling window and generates patterns with
 * time-based decay applied to confidence scores.
 *
 * Spec reference: navcaddy-engine.md R5
 */
@Singleton
class MissPatternAggregator @Inject constructor(
    private val repository: NavCaddyRepository,
    private val decayCalculator: PatternDecayCalculator
) {
    companion object {
        /** Default rolling window in days for pattern analysis */
        const val DEFAULT_WINDOW_DAYS = 30

        /** Default maximum number of shots to analyze */
        const val DEFAULT_SHOT_LIMIT = 50

        /** Minimum shots required to establish a pattern */
        const val MIN_SHOTS_FOR_PATTERN = 3

        /** Minimum frequency percentage to consider a direction a "pattern" (e.g., 0.3 = 30%) */
        const val MIN_FREQUENCY_THRESHOLD = 0.3f
    }

    /**
     * Aggregate patterns from recent shots.
     *
     * Analyzes shots within the rolling window (default 30 days or 50 shots, whichever is smaller)
     * and generates patterns with decay-adjusted confidence.
     *
     * @param days Rolling window in days (default 30)
     * @param maxShots Maximum number of shots to analyze (default 50)
     * @return List of aggregated patterns ordered by decayed confidence descending
     */
    suspend fun aggregatePatterns(
        days: Int = DEFAULT_WINDOW_DAYS,
        maxShots: Int = DEFAULT_SHOT_LIMIT
    ): List<MissPattern> {
        val shots = repository.getRecentShots(days)
            .first()
            .take(maxShots)

        return aggregatePatterns(shots)
    }

    /**
     * Aggregate patterns from a specific list of shots.
     *
     * Groups shots by miss direction and calculates frequency and confidence.
     * Applies time-based decay to confidence scores.
     *
     * @param shots List of shots to analyze
     * @return List of aggregated patterns ordered by decayed confidence descending
     */
    suspend fun aggregatePatterns(shots: List<Shot>): List<MissPattern> {
        if (shots.size < MIN_SHOTS_FOR_PATTERN) {
            return emptyList()
        }

        val currentTime = System.currentTimeMillis()

        // Filter out straight shots (we only care about misses)
        val missedShots = shots.filter {
            it.missDirection != null && it.missDirection != MissDirection.STRAIGHT
        }

        if (missedShots.isEmpty()) {
            return emptyList()
        }

        // Group by miss direction and calculate patterns
        return missedShots
            .groupBy { it.missDirection!! }
            .mapNotNull { (direction, directionShots) ->
                createPattern(
                    direction = direction,
                    shots = directionShots,
                    totalShots = shots.size,
                    currentTime = currentTime
                )
            }
            .sortedByDescending { it.confidence }
    }

    /**
     * Get patterns for a specific club.
     *
     * Analyzes shots for the club within the rolling window.
     *
     * @param clubId The club ID to filter by
     * @param days Rolling window in days (default 30)
     * @param maxShots Maximum number of shots to analyze (default 50)
     * @return List of club-specific patterns ordered by decayed confidence descending
     */
    suspend fun getPatternsForClub(
        clubId: String,
        days: Int = DEFAULT_WINDOW_DAYS,
        maxShots: Int = DEFAULT_SHOT_LIMIT
    ): List<MissPattern> {
        val shots = repository.getShotsByClub(clubId)
            .first()
            .filter { decayCalculator.isWithinRetentionWindow(it.timestamp, days) }
            .take(maxShots)

        val patterns = aggregatePatterns(shots)

        // Add club information to patterns
        return patterns.map { pattern ->
            if (shots.isNotEmpty()) {
                pattern.copy(club = shots.first().club)
            } else {
                pattern
            }
        }
    }

    /**
     * Get patterns for a specific pressure context.
     *
     * Analyzes pressure shots within the rolling window.
     *
     * @param days Rolling window in days (default 30)
     * @param maxShots Maximum number of shots to analyze (default 50)
     * @return List of pressure-specific patterns ordered by decayed confidence descending
     */
    suspend fun getPatternsForPressure(
        days: Int = DEFAULT_WINDOW_DAYS,
        maxShots: Int = DEFAULT_SHOT_LIMIT
    ): List<MissPattern> {
        val shots = repository.getShotsWithPressure()
            .first()
            .filter { decayCalculator.isWithinRetentionWindow(it.timestamp, days) }
            .take(maxShots)

        val patterns = aggregatePatterns(shots)

        // Mark patterns as pressure-related
        return patterns.map { pattern ->
            pattern.copy(
                pressureContext = PressureContext(
                    isUserTagged = shots.any { it.pressureContext.isUserTagged },
                    isInferred = shots.any { it.pressureContext.isInferred }
                )
            )
        }
    }

    /**
     * Get patterns for a specific club under pressure.
     *
     * Analyzes pressure shots for the club within the rolling window.
     *
     * @param clubId The club ID to filter by
     * @param days Rolling window in days (default 30)
     * @param maxShots Maximum number of shots to analyze (default 50)
     * @return List of club-pressure patterns ordered by decayed confidence descending
     */
    suspend fun getPatternsForClubAndPressure(
        clubId: String,
        days: Int = DEFAULT_WINDOW_DAYS,
        maxShots: Int = DEFAULT_SHOT_LIMIT
    ): List<MissPattern> {
        val shots = repository.getShotsByClub(clubId)
            .first()
            .filter { it.pressureContext.hasPressure }
            .filter { decayCalculator.isWithinRetentionWindow(it.timestamp, days) }
            .take(maxShots)

        val patterns = aggregatePatterns(shots)

        // Add club and pressure information
        return patterns.map { pattern ->
            pattern.copy(
                club = if (shots.isNotEmpty()) shots.first().club else null,
                pressureContext = PressureContext(
                    isUserTagged = shots.any { it.pressureContext.isUserTagged },
                    isInferred = shots.any { it.pressureContext.isInferred }
                )
            )
        }
    }

    /**
     * Get patterns from repository with decay applied.
     *
     * Returns persisted patterns with current decayed confidence values.
     *
     * @return Flow of patterns with decay applied
     */
    fun getPatternsWithDecay(): Flow<List<MissPattern>> {
        val currentTime = System.currentTimeMillis()
        return repository.getMissPatterns().map { patterns ->
            patterns
                .map { pattern ->
                    pattern.copy(
                        confidence = decayCalculator.calculateDecayedConfidence(
                            baseConfidence = pattern.confidence,
                            lastOccurrence = pattern.lastOccurrence,
                            currentTime = currentTime
                        )
                    )
                }
                .filter { it.confidence > 0.01f } // Filter out nearly-decayed patterns
                .sortedByDescending { it.confidence }
        }
    }

    /**
     * Create a miss pattern from a group of shots with the same direction.
     *
     * Calculates frequency, base confidence, and applies time-based decay.
     *
     * @param direction The miss direction for this pattern
     * @param shots Shots with this miss direction
     * @param totalShots Total number of shots in the analysis window
     * @param currentTime Current timestamp for decay calculation
     * @return MissPattern or null if pattern doesn't meet thresholds
     */
    private fun createPattern(
        direction: MissDirection,
        shots: List<Shot>,
        totalShots: Int,
        currentTime: Long
    ): MissPattern? {
        val frequency = shots.size
        val frequencyRatio = frequency.toFloat() / totalShots

        // Filter out patterns below threshold
        if (frequencyRatio < MIN_FREQUENCY_THRESHOLD) {
            return null
        }

        // Calculate base confidence from frequency ratio
        val baseConfidence = frequencyRatio.coerceIn(0f, 1f)

        // Find most recent occurrence
        val lastOccurrence = shots.maxOfOrNull { it.timestamp } ?: currentTime

        // Apply decay to base confidence
        val decayedConfidence = decayCalculator.calculateDecayedConfidence(
            baseConfidence = baseConfidence,
            lastOccurrence = lastOccurrence,
            currentTime = currentTime
        )

        return MissPattern(
            id = UUID.randomUUID().toString(),
            direction = direction,
            club = null, // Club will be added by caller if applicable
            frequency = frequency,
            confidence = decayedConfidence,
            pressureContext = null, // Pressure context will be added by caller if applicable
            lastOccurrence = lastOccurrence
        )
    }

    /**
     * Save aggregated patterns to the database.
     *
     * Persists patterns for retrieval without re-aggregation.
     *
     * @param patterns List of patterns to save
     */
    suspend fun savePatterns(patterns: List<MissPattern>) {
        patterns.forEach { pattern ->
            repository.updatePattern(pattern)
        }
    }

    /**
     * Clear stale patterns from the database.
     *
     * Removes patterns outside the retention window.
     *
     * @return Number of patterns deleted
     */
    suspend fun clearStalePatterns(): Int {
        return repository.deleteStalePatterns()
    }
}
